cclib/parser/gaussianparser.py

```diff
@@ -1853,46 +1871,73 @@ def natural_orbital_single_spin_parsing(inputfile, updateprogress_title):
         #     2  C    0.002063
         # ...
         #
-        if line[1:25] == "Mulliken atomic charges:" or line[1:18] == "Mulliken charges:" or \
-           line[1:23] == "Lowdin Atomic Charges:" or line[1:16] == "Lowdin charges:" or \
-           line[1:37] == "Mulliken charges and spin densities:" or \
-           line[1:32] == "Mulliken atomic spin densities:":
-
-            has_spin = 'spin densities' in line 
-            has_charges = 'charges' in line 
-
-            if has_charges and not hasattr(self, "atomcharges"):
-                self.atomcharges = {}
-
-            if has_spin and not hasattr(self, "atomspins"):
-                self.atomspins = {}
-
-            ones = next(inputfile)
-
-            charges = []
-            spins = []
-            nline = next(inputfile)
-            while not "Sum of" in nline:
-                if has_charges:
-                    charges.append(float(nline.split()[2]))
-
-                if has_spin and has_charges:
-                    spins.append(float(nline.split()[3]))
-
-                if has_spin and not has_charges:
-                    spins.append(float(nline.split()[2]))
-
-                nline = next(inputfile)
-
-            if "Mulliken" in line:
-                if has_charges:
-                    self.atomcharges["mulliken"] = charges
-                if has_spin:
-                    self.atomspins["mulliken"] = spins
-
-            elif "Lowdin" in line:
-                self.atomcharges["lowdin"] = charges
-
+        # Spins may be included in the same section, e.g.
+        #
+        # Mulliken charges and spin densities:
+        #               1          2
+        #     1  O   -0.596188   0.019133
+        #     2  C    0.320624   0.000869
+        #
+        # APT and Lowdin charges are also displayed in this way
+        def extract_charges_spins(property):
+            """Extracts atomic charges and spin densities into self.atomcharges and self.atomspins dictionaries
+    
+            Inputs:
+                property - property type to be extracted as a string(e.g. Mulliken, Lowdin, APT)
+            """
+
+            headers = [" atomic charges:",
```

Comment:
Not specific to this line, but a general design comment and question for everyone. Right now, the code is written like
``` python
def extract(self, inputfile, line):
    def extract_charges_spins(prop):
        # do discovery and parsing here
    extract_charges_spins("mulliken")
    extract_charges_spins("lowdin")
    extract_charges_spins("APT")
```
where in other places we have code that looks more like
``` python
def extract_charges_spins(self, intputfile, line, prop):
    # do parsing here
def extract(self, inputfile, line):
    # do discovery here
    if discovered_header_for_prop in line:
        prop = ...
        extract_charges_spins(self, inputfile, line, prop)
```
I think I like your approach more, since the latter misleads you into thinking that the method is usable outside of `extract`, which it is, but not practically, since for expected behavior to happen, `line` really needs to be in the right position before the method is called. My main confusion comes from hiding the detection (`headers` and related) inside the function. It's not clear if rewriting this bit would require removing too much from the function. What do you think?

Reply:
Personally, I like the extract_charges_spins() within extract(), and next to the related code, just because I find it tidier and more readable when it's all together. However, if the extract_charges_spins function was being used elsewhere, or was intended to be used outside of extract (which, as you say, in theory it can be, but in reality it won't be), then that would be a different story. Depending on the consunsus, I'm happy to change things about.

I agree with regards to the headers and detection part (I probably just got carried away trying to pack everything inside one function). In hindsight, I think it would make more sense (and be fine) if the headers/line-finding aspect came outside the function. For example, something like this might be better:

```
def extract_charges_spins():
    # parsing

props = ["mulliken", "APT", "Lowdin"]
headers = [" atomic charges:", " charges:", etc]

if hasattr(self, "natom") and hasattr(self, "nhydrogen"):
    for prop in props:
          for header in headers:
                 if '{}{}'.format(prop,header).lower() in line.lower():
                          extract_charges_spins()
```

I can rewrite that if that sounds better.

Reply:
Yes, I think that's the best solution.